vnext/ReactWindowsCore/WinRTWebSocketResource.h

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// clang-format off
+
+#pragma once
+
+//#include <Mso/eventWaitHandle/eventWaitHandle.h>
+#include <winrt/Windows.Networking.Sockets.h>
+#include <winrt/Windows.Storage.Streams.h>
+#include <IWebSocketResource.h>
+
+#include <future>
+#include <queue>
+
+namespace Microsoft::React
+{
+
+class WinRTWebSocketResource : public IWebSocketResource
+{
+  winrt::Windows::Foundation::Uri m_uri;
+  winrt::Windows::Networking::Sockets::MessageWebSocket m_socket;
+  winrt::Windows::Networking::Sockets::MessageWebSocket::MessageReceived_revoker m_revoker;
+  winrt::Windows::Storage::Streams::DataWriter m_writer;
+
+  std::atomic_bool m_connectRequested;
+  std::atomic_bool m_sendRequested;
+  // Mso::ManualResetEvent m_connectPerformed;
+  // Mso::ManualResetEvent m_sendPerformed;
+  std::promise<void> m_connectPerformed;
+
+  CloseCode m_closeCode;
+  std::string m_closeReason;
+  std::queue<std::pair<std::string, bool>> m_writeQueue;
+
+  std::function<void()> m_connectHandler;
+  std::function<void()> m_pingHandler;
+  std::function<void(std::size_t)> m_writeHandler;
+  std::function<void(std::size_t, const std::string&)> m_readHandler;
+  std::function<void(CloseCode, const std::string&)> m_closeHandler;
+  std::function<void(Error&&)> m_errorHandler;
+
+  WinRTWebSocketResource(winrt::Windows::Foundation::Uri&& uri);
+  ~WinRTWebSocketResource() override;
+
+  winrt::fire_and_forget PerformConnect();
+  winrt::fire_and_forget PerformPing();
+  winrt::fire_and_forget PerformWrite();
+  winrt::fire_and_forget PerformClose();
+
+  void OnMessageReceived(
+    winrt::Windows::Networking::Sockets::IWebSocket const& sender,
+    winrt::Windows::Networking::Sockets::MessageWebSocketMessageReceivedEventArgs const& args);
+  void OnClosed(
+    winrt::Windows::Networking::Sockets::IWebSocket const& sender,
+    winrt::Windows::Networking::Sockets::WebSocketClosedEventArgs const& args);
+
+  void Stop();
+
+public:
```

Comment:
Nit: could we put the public interface first?

Reply:
Could. Keeping it as is for now.
Open to refactoring later after getting team consensus.

Reply:
In general sequence should be:
1. Public Constructors
2. Public Methods
3. Private Methods
4. Private Member Variables

---
In reply to: [394800738](https://github.com/microsoft/react-native-windows/pull/4330#discussion_r394800738) [](ancestors = 394800738)